velox/functions/lib/aggregates/noisy_aggregation/sketch/SfmSketch.h

```diff
@@ -0,0 +1,253 @@
+  }
+
+  // Make a non-private sketch private.
+  void enablePrivacy(double epsilon);
+
+  // Get a default randomization strategy.
+  static SecureRandomizationStrategy getDefaultRandomizationStrategy() {
+    return SecureRandomizationStrategy();
+  }
+
+  // Make a non-private sketch private with a randomization strategy.
+  template <typename RandomizationStrategy>
+  void enablePrivacy(
+      double epsilon,
+      RandomizationStrategy randomizationStrategy) {
+    VELOX_CHECK_NOT_NULL(
+        &randomizationStrategy, "randomizationStrategy can't be null.");
+    VELOX_CHECK(!privacyEnabled(), "privacy is already enabled.");
+    VELOX_CHECK(
+        epsilon > 0,
+        "Epsilon must be greater than zero or equal to kNonPrivateEpsilon");
+    randomizedResponseProbability_ =
+        calculateRandomizedResponseProbability(epsilon);
+
+    // Toggle each bit with a securely generated probability.
+    for (uint32_t i = 0; i < numberOfBits(); ++i) {
+      if (randomizationStrategy.nextBoolean(randomizedResponseProbability_)) {
+        bits::negateBit(bits_.data(), static_cast<uint64_t>(i));
+      }
+    }
+  }
+
+  // Get the number of bits that are set to 1.
+  int32_t countBits() const;
+
+  // Estimate the cardinality of the sketch via maximum psuedolikelihood
+  // (Newton's method).
+  uint64_t cardinality() const;
+
+  // Size of the sketch in bytes
+  size_t serializedSize() const;
+
+  // Serialize the sketch into a char array.
+  void serialize(char* out) const;
+
+  // Deserialize the sketch from a char array.
+  static SfmSketch deserialize(const char* in, HashStringAllocator* allocator);
+
+  // Probability of a 1-bit remaining a 1-bit under randomized response.
+  double onProbability() const {
+    return 1 - randomizedResponseProbability_;
+  }
+
+  uint32_t numberOfBits() const {
+    return static_cast<uint32_t>(bits_.size() * 8);
+  }
+
+  double randomizedResponseProbability() const {
+    return randomizedResponseProbability_;
+  }
+
+  uint32_t indexBitLength() const {
+    return indexBitLength_;
+  }
+
+  uint32_t precision() const {
+    return precision_;
+  }
+
+  std::vector<int8_t, facebook::velox::StlAllocator<int8_t>> bits() const {
+    return bits_;
+  }
+
+  bool privacyEnabled() const {
+    return randomizedResponseProbability_ > 0;
+  }
+
+ private:
+  // Epsillon for non-private sketch.
+  static constexpr double kNonPrivateEpsilon =
+      std::numeric_limits<double>::infinity();
+
+  // Maximum number of iterations for Newton's method.
+  static constexpr int32_t kMaxIteration = 1000;
+
+  // Just a tag used to be compatible with the old Java implementation.
+  static constexpr int8_t kFormatTag = 7;
+
+  // Count the number of trailing zeros in a 64-bits hash value starting from
+  // the most significant bit after the 'indexBitLength' bits.
+  static uint32_t numberOfTrailingZeros(uint64_t hash, uint32_t indexBitLength);
```

Comment:
This is only used in .cpp, hence, can be moved these. Smaller header files => faster builds.